conans/client/remote_registry.py

```diff
@@ -83,68 +85,88 @@ def __init__(self, filename, lockfile, output):
+        save(self._filename, self._contents)
 
 
-class _ReferencesRegistry(_Registry):
+class _GenericReferencesRegistry(_Registry):
 
     @property
     def list(self):
         with fasteners.InterProcessLock(self._lockfile, logger=logger):
-            _, refs = self._load()
-            return refs
+            _, refs, prefs = self._load()
+            return self._select_refs(refs, prefs)
 
-    def remove(self, conan_reference, quiet=False):
-        assert(isinstance(conan_reference, ConanFileReference))
+    def remove(self, ref, quiet=False):
+        assert(isinstance(ref, ConanFileReference) or isinstance(ref, PackageReference))
         with fasteners.InterProcessLock(self._lockfile, logger=logger):
-            remotes, refs = self._load()
+            remotes, rrefs, prefs = self._load()
+            refs = self._select_refs(rrefs, prefs)
             try:
-                del refs[str(conan_reference)]
-                self._save(remotes, refs)
+                del refs[str(ref)]
+                self._save(remotes, rrefs, prefs)
             except:
                 if not quiet:
-                    self._output.warn("Couldn't delete '%s' from remote registry"
-                                      % str(conan_reference))
+                    self._output.warn("Couldn't delete '%s' from remote registry" % str(ref))
 
-    def get(self, conan_reference):
-        assert(isinstance(conan_reference, ConanFileReference))
+    def get(self, ref):
+        assert(isinstance(ref, ConanFileReference) or isinstance(ref, PackageReference))
         with fasteners.InterProcessLock(self._lockfile, logger=logger):
-            remotes, refs = self._load()
-            remote_name = refs.get(str(conan_reference))
+            remotes, rrefs, prefs = self._load()
+            refs = self._select_refs(rrefs, prefs)
+            remote_name = refs.get(str(ref))
             try:
                 return Remote(remote_name, remotes[remote_name][0], remotes[remote_name][1])
             except KeyError:
                 return None
 
-    def set(self, conan_reference, remote_name, check_exists=False):
-        assert(isinstance(conan_reference, ConanFileReference))
+    def set(self, ref, remote_name, check_exists=False):
+        assert(isinstance(ref, ConanFileReference) or isinstance(ref, PackageReference))
         with fasteners.InterProcessLock(self._lockfile, logger=logger):
-            remotes, refs = self._load()
+            remotes, rrefs, prefs = self._load()
+            refs = self._select_refs(rrefs, prefs)
             if check_exists:
-                if conan_reference in refs:
-                    raise ConanException("%s already exists. Use update" % conan_reference)
+                if ref in refs:
+                    raise ConanException("%s already exists. Use update" % ref)
                 if remote_name not in remotes:
                     raise ConanException("%s not in remotes" % remote_name)
```

Comment:
This check should be done outside the ``check_exists`` condition, as it is done in the ``update`` method too.